TASK-33: Set Event Location Backend

RGPosadas opened the pull request:
This PR closes #124 

In addition to the task's AC, we also took the liberty to:
* fix some bugs 
* cleanup the styling with the location autocomplete
* Added more types for better autocomplete (ICreateEventForm, etc.)
* Fix some of the placeholders that were in the frontend code for location, since now locations are being saved and fetched when events are fetched.
* Move LocationAutocomplete to be a REST endpoint instead of a graphql resolver. This was mainly done for simplicity
* Remove the axiosConfig file, in favour of mocking the axios itself during tests

What to look out for:
* On the frontend: event cards and event previews now have the correct location title as that data is being fetched from the backend.
* On the backend: creating an event also fills in the Location table. If Google Autocomplete is used, then it will save the placeId in the Location table. If Current Location is chosen, then coordinates will also be saved in the Point table.

NOTE: If you have seeded your database, you might run into some issues, since the locationId for new events is set to null by the seeder. Please wipe your db and create new events through the interface when testing this PR (or fix the locationid column in your db event table)

RGPosadas committed:
feat(mull-api): setup endpoints for getting Location

Created endpoints for fetching and saving locations. Also updated/added entities related to
Location.

BREAKING CHANGE: Affects the Event database table

Refs #124

Co-authored-by: Cristian Aldea <aldea.cristian@outlook.com>
Co-authored-by: Jonathan Hsu <bit172@gmail.com>

RGPosadas committed:
refactor(location): move LocationAutocomplete to a REST endpoint

Location can now also be sent through Create Event to the backend. Also added ToastOptions to useToast.

With the GraphQL endpoint, in order to use IGooglePlace as a return type for the autocompleteLocations request, we needed to create a new class called GooglePlace annotated with the decorator @ObjectType. This ended up being convoluted
for no reason, therefore we ended up moving everything to a REST endpoint.

Refs #124

Co-authored-by: Cristian Aldea <aldea.cristian@outlook.com>

cristian-aldea committed:
fix(location auto): Fix bug where api was called for current location

Also, fixed other small issues

ref #124

cristian-aldea committed:
fix(location): Add types and fix bugs

Refs #124

RGPosadas committed:
refactor(mull-ui): add/update fragments to simplify frontend queries

Refs #124

Co-authored-by: Cristian Aldea <aldea.cristian@outlook.com>

RGPosadas committed:
refactor(mull-ui): cleanup event.location.title placeholders

Refs #124

Co-authored-by: Cristian Aldea <aldea.cristian@outlook.com>

RGPosadas committed:
refactor(mull-ui): fix inline styling and move them to scss file

Refs #124

RGPosadas committed:
refactor(mull-api): refactor Event-Location to have a 1-1 relationship

We needed to have a 1-1 relationship for Event-Location as we weren't taking advantage of the Many-1
relationship they originally had. This also made it easier to manipulate the relations using
QueryBuilder.

Refs #124

Co-authored-by: Cristian Aldea <aldea.cristian@outlook.com>

RGPosadas committed:
feat(mull-api): return location along with all event queries

Refs #124

Co-authored-by: Cristian Aldea <aldea.cristian@outlook.com>

RGPosadas committed:
refactor(locationautocomplete): remove GraphQL endpoint for Location Autocomplete

Refs #124

RGPosadas committed:
chore(jest.config): do not run a report on generated files and on static assets

Refs #124

RGPosadas committed:
test(mull-ui): update snapshots

Refs #124

RGPosadas committed:
test(mull-ui): add location to frontend unit tests

Refs #124

cristian-aldea committed:
fix(test): Fix app controller tests

ref #124

cristian-aldea committed:
fix(test): Fix Location autocomplete tests

ref #124

cristian-aldea committed:
fix(lint): fix linting

Refs #124

RGPosadas committed:
test(event.service): add location to event.service tests

Refs #124

RGPosadas committed:
test(mull-api): add unit tests for event.service and event.resolver

Refs #124

RGPosadas committed:
test(e2e): fix failing e2e tests

Refs #124

cristian-aldea commented on apps/mull-api/src/app/entities/event.entity.ts:
```diff
@@ -76,9 +76,10 @@ export class Event implements IEvent {
-    /* istanbul ignore next */ () => Location,
-    /* istanbul ignore next */ (location) => location.events
-  )
+  @Field({ nullable: true })
+  @JoinColumn()
+  @OneToOne(/* istanbul ignore next */ () => Location, (location) => location.event, {
+    cascade: true,
+  })
```
This was done since it would simplify some of the queries we needed to change, and since we're currently not taking advantage of this many to one relationship (we're not querying for locations when creating an event to avoid duplicates, etc.)

cristian-aldea commented on apps/mull-ui/jest.config.js:
```diff
@@ -8,4 +8,5 @@ module.exports = {
   coverageDirectory: '../../coverage/apps/mull-ui',
   setupFiles: ['./jest.config.setup.js'],
   displayName: 'mull-ui',
+  coveragePathIgnorePatterns: ['generated', 'assets'],
```
Excluding the `graphql.tsx` and icons from coverage.

cristian-aldea commented on apps/mull-ui/src/app/pages/create-event/location-autocomplete/location-autocomplete-textbox.tsx:
```diff
@@ -3,46 +3,66 @@ import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+          .get(encodeURI(`${environment.backendUrl}/api/location-autocomplete?search=${search}`))
+          .then((data) => data.data)
+          .then((data: IGooglePlace[]) => {
+            setOptions([CURRENT_LOCATION, ...data]);
```
This is so that users can still use the `Current Location` option after they started to type

r-saba replied:
Nice

cristian-aldea committed:
refactor(location): Cleanup code and remove axios config

Axios config was only used during tests to be mocked. The usage of that instance was replaced with a use of axios itself, and the mocks were adjusted too. So there is no longer need to keep the axios config.

ref #124

bit172 commented on apps/mull-ui/src/graphql/queries.gql:
```diff
@@ -1,73 +1,53 @@
-query Event($id: Int!) {
-  event(id: $id) {
-    id
+fragment EventPageContent on Event {
```
Nice use of fragments! 🚀

bit172 reviewed (approved):
Good Work! The only thing I could say is that the seeder will no longer be compatible with our application, but that could be done in another ticket, or perhaps do we need to add new criteria for implementing features 🤔. Anyway, this is probably a topic for a meeting. 🚀

**Acceptance Criteria**
- [x] Connects with Location frontend #29 
- [x] Location is stored when an event is created 
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function

cristian-aldea committed:
fix(sonarcloud): Remove redundant conditional

ref #124

JimLau49 commented on apps/mull-ui/src/generated/graphql.tsx:
```diff
@@ -663,21 +659,13 @@ export type DiscoverEventsQueryResult = Apollo.QueryResult<DiscoverEventsQuery,
-    description
-    startDate
-    endDate
+    ...EventCardContent
```
nice :)

TheGreatMarkus commented:
@bit172 

I think that should be addressed as a separate ticket, since it wasn't part of the AC for this ticket, since this ticket was drafted before the seeding that done. Definitely something to talk about for future tickets.

r-saba commented on apps/mull-api/src/app/event/inputs/event.input.ts:
```diff
@@ -27,8 +28,8 @@ export class CreateEventInput implements Partial<Event> {
-  // @Field()
-  // location: Location;
+  @Field(/* istanbul ignore next*/ () => LocationInput)
+  location?: Location;
```
```suggestion
  location: Location;
```

r-saba commented on apps/mull-api/src/app/entities/location.entity.ts:
```diff
@@ -1,15 +1,44 @@
+  id: number;
+
+  @Field()
+  @Column('double')
```
Maybe we could use the `Decimal` type over `double` here.

RGPosadas replied:
You're right, it's more appropriate for lat and long.

r-saba commented on apps/mull-api/src/app/location/location.resolver.ts:
```diff
@@ -0,0 +1,19 @@
+import { LocationInput } from './inputs/location.input';
+import { LocationService } from './location.service';
+
+@Resolver('Location')
```
```suggestion
@Resolver(() => Location)
```

r-saba commented on apps/mull-ui/src/app/components/event-card/event-card.tsx:
```diff
@@ -58,10 +61,7 @@ export const EventCard = ({ event, style = {}, onClick, isJoined = false }: Even
-            {/* TODO: Dynamically add address. Currently not available in the data from the query*/}
-            15km • 1455 Boulevard de Maisonneuve O, Montréal, QC H3G 1M8
-          </div>
+          <div className="event-card-location">{`${distance}km • ${event.location?.title}`}</div>
```
```suggestion
          <div className="event-card-location">{`${distance}km • ${event.location.title}`}</div>
```

r-saba reviewed (changes requested):
LGTM just a few comments you can address when you have time

RGPosadas committed:
refactor(mull-api): make Location necessary for CreateEventInput, and optional for UpdateEventInput

Refs #124

RGPosadas committed:
refactor(event-card): make event.location.title mandatory

Refs #124

RGPosadas committed:
refactor(mull-api): set lat/long datatype to DECIMAL

Also use anonymous funcs instead of strings for Resolver decorators.

Refs #124

JimLau49 reviewed (approved):
Working good for me no other suggestions from me
**Acceptance Criteria**
- [x] Connects with Location frontend #29 
- [x] Location is stored when an event is created 
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function

r-saba reviewed (approved):
LGTM!
**Acceptance Criteria**
- [X] Connects with Location frontend #29 
- [X] Location is stored when an event is created 
- [X] Documentation present where necessary
- [X] Unit and UI tests are implemented for every component and function
